weasyl/controllers/two_factor_auth.py

```diff
@@ -119,21 +115,26 @@ def tfa_init_qrcode_get_(request):
 def tfa_init_qrcode_post_(request):
     # Strip any spaces from the TOTP code (some authenticators display the digits like '123 456')
     tfaresponse = request.params['tfaresponse'].replace(' ', '')
-    tfa_secret_sess = _get_totp_code_from_session()
+    tfa_secret = _retrieve_information_from_session(key="tfa_secret")
+    print(tfa_secret)
+
+    # TODO: Implement
+    #_store_information_on_session(key="authenticator_comment", value=request.params['comment'])
 
     # Check to see if the tfaresponse matches the tfasecret when run through the TOTP algorithm
-    tfa_secret, recovery_codes = tfa.init_verify_tfa(request.userid, tfa_secret_sess, tfaresponse)
+    tfa_secret, recovery_codes = tfa.init_verify_tfa(request.userid, tfa_secret, tfaresponse)
```

Comment:
This is the reason it's failing when an incorrect TOTP code is being passed here. ``(False, None)`` is being returned, and the template is trying to ``__getitem__`` on the Boolean.

Suggested fix:
```python
    # Check to see if the tfaresponse matches the tfasecret when run through the TOTP algorithm
   recovery_codes = tfa.init_verify_tfa(request.userid, tfa_secret, tfaresponse)

    # The 2FA TOTP code did not match with the generated 2FA secret
    if not recovery_codes:
        return Response(define.webpage(request.userid, "control/2fa/qrcode.html", [
            define.get_display_name(request.userid),
            tfa_secret,
            tfa.generate_tfa_qrcode(request.userid, tfa_secret),
            "2fa",
            False,
], title="Enable 2FA: Step 2"))
```

We already store ``tfa_secret`` on the session itself, so there's no reason to redefine it. We can simply refactor ``tfa.init_verify_tfa`` to either return a list of recovery codes, or ``None``, which still satisfies the need to verify that we correctly validated the TOTP token generated by the authenticator. Since this current commit doesn't have that function being modified...

In ``/weasyl/two_factor_auth.py``...
```python
def init_verify_tfa(userid, tfa_secret, tfa_response):
    """
    Verify that the user has successfully set-up 2FA in Google Authenticator
    (or similar), and generate recovery codes for the user.
    This function is part one of two in enabling 2FA. Successful verification of this phase
    ensures that the user's authentication app is working correctly.
    Parameters:
        userid: The userid of the calling user.
        tfa_secret: The 2FA secret generated from tfa_init(); retrieved from the
        verification page's form information.
        tfa_response: The 2FA challenge-response code to verify against tfa_secret.
    Returns: A list of recovery codes as generated by `generate_recovery_codes()` if ``tfa_response`` corresponds to ``tfa_secret``, otherwise ``None``.
    """
    totp = pyotp.TOTP(tfa_secret)
    # If the provided `tfa_response` matches the TOTP value, add the value and return recovery codes
    if totp.verify(tfa_response):
        return generate_recovery_codes()
    else:
        None
```